src/util/pointer_expr.h

```diff
@@ -381,6 +381,235 @@ inline address_of_exprt &to_address_of_expr(exprt &expr)
   return ret;
 }
 
+/// \brief Operator to return the address of an object
+class object_address_exprt : public nullary_exprt
+{
+public:
+  explicit object_address_exprt(const symbol_exprt &);
+
+  irep_idt object_identifier() const
+  {
+    return get(ID_identifier);
+  }
+
+  const pointer_typet &type() const
+  {
+    return static_cast<const pointer_typet &>(exprt::type());
+  }
+
+  pointer_typet &type()
+  {
+    return static_cast<pointer_typet &>(exprt::type());
+  }
+
+  /// returns the type of the object whose address is represented
+  const typet &object_type() const
+  {
+    return type().subtype();
+  }
+
+  symbol_exprt object_expr() const;
+};
+
+template <>
+inline bool can_cast_expr<object_address_exprt>(const exprt &base)
+{
+  return base.id() == ID_object_address;
+}
+
+inline void validate_expr(const object_address_exprt &value)
+{
+  validate_operands(value, 1, "object_address must have one operand");
+}
+
+/// \brief Cast an exprt to an \ref object_address_exprt
+///
+/// \a expr must be known to be \ref object_address_exprt.
+///
+/// \param expr: Source expression
+/// \return Object of type \ref object_address_exprt
+inline const object_address_exprt &to_object_address_expr(const exprt &expr)
+{
+  PRECONDITION(expr.id() == ID_object_address);
+  const object_address_exprt &ret =
+    static_cast<const object_address_exprt &>(expr);
+  validate_expr(ret);
+  return ret;
+}
+
+/// \copydoc to_object_address_expr(const exprt &)
+inline object_address_exprt &to_object_address_expr(exprt &expr)
+{
+  PRECONDITION(expr.id() == ID_object_address);
+  object_address_exprt &ret = static_cast<object_address_exprt &>(expr);
+  validate_expr(ret);
+  return ret;
+}
+
+/// \brief Operator to return the address of a field relative
+/// to a base address
+class field_address_exprt : public unary_exprt
+{
+public:
+  /// constructor for field addresses.
+  /// The base address must be a pointer to a compound type.
```

Comment:
So should the `base` actually be an `object_address_exprt`?

Reply:
No, this can be any pointer, say "NULL", or a pointer-typed symbol.